### PR TITLE
Upgrade to PNPM 4.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

**Comment by iclanton** [comment]
@sachinjoseph - does this look right?

**Comment by octogonz** [comment]
@zkochan :-)

**Comment by zkochan** [comment]
Maybe off topic but were the ram usage issues solved by pnpm v4? Cc @sachinjoseph

**Comment by apostolisms** [comment]
@zkochan I think we are still having these issues. We haven't tested 4.1.x though

File: common/config/rush/pnpm-lock.yaml [hunk 1]
```diff
@@ -3,15 +3,15 @@ dependencies:
...
+  '@pnpm/link-bins': 1.0.3_@pnpm+logger@1.0.2
   '@pnpm/logger': 1.0.2
   '@rush-temp/api-documenter': 'file:projects/api-documenter.tgz'
   '@rush-temp/api-documenter-test': 'file:projects/api-documenter-test.tgz'
   '@rush-temp/api-extractor': 'file:projects/api-extractor.tgz'
   '@rush-temp/api-extractor-lib1-test': 'file:projects/api-extractor-lib1-test.tgz'
   '@rush-temp/api-extractor-lib2-test': 'file:projects/api-extractor-lib2-test.tgz'
   '@rush-temp/api-extractor-lib3-test': 'file:projects/api-extractor-lib3-test.tgz'
-  '@rush-temp/api-extractor-model': 'file:projects/api-extractor-model.tgz'
+  '@rush-temp/api-extractor-model': 'file:projects/api-extractor-model.tgz_tslint@5.12.1'
```
Comment (iclanton) [inline common/config/rush/pnpm-lock.yaml]: @zkochan - this is the only thing that's concerning. Does this seem right?

Comment (zkochan) [reply common/config/rush/pnpm-lock.yaml]: yes. This is correct. It happens when the tarball has a peer dependnency on tslint

**Comment by octogonz** [comment]
@iclanton @sachinjoseph  Could you test whether publishing is broken with PNPM 4.x?

When I tried this in the TSDoc repo, I got this error:
```
* EXECUTING: D:\a\1\s\common\temp\pnpm-local\node_modules\.bin\pnpm publish --//registry.npmjs.org/:_authToken=*** (eslint-plugin)
 ERROR  Unknown option '//registry.npmjs.org/:_authToken'
For help, run: pnpm help publish
```

Seems like `--//registry.npmjs.org/:_authToken` is no longer allowed as a command-line option. (?)

As soon as PR https://github.com/microsoft/rushstack/pull/1602 is merged, we will recommend not to use this approach.  But we don't want it to be broken.

**Comment by octogonz** [comment]
Also, it's probably time to bump the `rush init` versions in [assets/rush-init/rush.json](https://github.com/microsoft/rushstack/blob/e125ce4a1aa62f2808a1ecf542fa7b0748498c6d/apps/rush-lib/assets/rush-init/rush.json#L29)...

**Comment by octogonz** [comment]
> Seems like `--//registry.npmjs.org/:_authToken` is no longer allowed as a command-line option. (?)

[PNPM Gitter discussion](https://gitter.im/pnpm/pnpm?at=5dc29291a03ae1584f788030) about this problem. @zkochan is proposing to reintroduce this in PNPM 4 but with a different command line syntax.

**Comment by zkochan** [comment]
I was lazy to implement new syntax, so I allowed the old one. It should work with v4.1.8

**Comment by octogonz** [comment]
@iclanton @apostolisms Are there any remaining issues blocking this PR?

**Comment by iclanton** [comment]
This has already been checked in [here](https://github.com/microsoft/rushstack/pull/1732)

**Pull request closed without merging**